Plugins processes not cleaned up, possibly only when errors occur

**Describe the bug**
Lyra is not cleaning up plugin processes on all occasions and I am finding this sort of thing in my process list:

```
  501 81143     1   0 11Feb19 ??         0:00.02 /System/Library/PrivateFrameworks/KerberosHelper/Helpers/DiskUnmountWatcher
  501 83702     1   0  4Feb19 ??         0:21.65 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build751753963/b001/exe/main --debug plugin example
  501 83703     1   0  4Feb19 ??         0:23.97 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build751753963/b001/exe/main --debug plugin identity
  501 83704     1   0  4Feb19 ??         0:24.55 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build751753963/b001/exe/main --debug plugin aws
  501 83705     1   0  4Feb19 ??         1:02.73 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build751753963/b001/exe/main --debug plugin tfaws
  501 83708     1   0  4Feb19 ??         0:38.76 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build751753963/b001/exe/main --debug plugin tfkubernetes
  501 87235     1   0  4Feb19 ??         0:23.31 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build695187835/b001/exe/main --debug plugin example
  501 87236     1   0  4Feb19 ??         0:24.09 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build695187835/b001/exe/main --debug plugin identity
  501 87237     1   0  4Feb19 ??         0:25.16 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build695187835/b001/exe/main --debug plugin aws
  501 87238     1   0  4Feb19 ??         1:03.76 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build695187835/b001/exe/main --debug plugin tfaws
  501 87242     1   0  4Feb19 ??         0:39.54 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build695187835/b001/exe/main --debug plugin tfkubernetes
  501 87982     1   0  4Feb19 ??         0:23.01 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build002614375/b001/exe/main --debug plugin example
  501 87983     1   0  4Feb19 ??         0:22.48 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build002614375/b001/exe/main --debug plugin identity
  501 87984     1   0  4Feb19 ??         0:25.68 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build002614375/b001/exe/main --debug plugin aws
  501 87987     1   0  4Feb19 ??         0:38.82 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build002614375/b001/exe/main --debug plugin tfkubernetes
  501 87989     1   0  4Feb19 ??         1:45.03 /var/folders/26/7dm8dk111n1109jy0p7t0wgr0000gn/T/go-build002614375/b001/exe/main --debug plugin puppet
```

It might only be in the case that the run fails.

**To Reproduce**
Not sure, but I have recently been testing both plugins that panic and have had plenty of badly formed workflows, so could be at either stage of the run.

**Expected behavior**
No plugin processes left alive once Lyra exits.

Looks like this has cropped up again.

I'm seeing many plugins running after failed Lyra runs:
```
  501 42458     1   0 13Mar19 ??         0:00.27 build/goplugin-example
  501 42459     1   0 13Mar19 ??         0:00.27 build/goplugin-identity
  501 42460     1   0 13Mar19 ??         0:00.26 build/goplugin-puppetdsl
  501 42461     1   0 13Mar19 ??         1:20.24 build/goplugin-tf-aws
  501 42463     1   0 13Mar19 ??         0:46.19 build/goplugin-tf-azurerm
  501 42464     1   0 13Mar19 ??         0:10.92 build/goplugin-tf-github
  501 42465     1   0 13Mar19 ??         0:39.67 build/goplugin-tf-google
  501 42466     1   0 13Mar19 ??         0:42.05 build/goplugin-tf-kubernetes
  501 42500     1   0 13Mar19 ??         0:11.81 build/goplugin-aws
  501 42501     1   0 13Mar19 ??         0:00.25 build/goplugin-example
  501 42502     1   0 13Mar19 ??         0:00.25 build/goplugin-identity
  501 42503     1   0 13Mar19 ??         0:00.25 build/goplugin-puppetdsl
  501 42504     1   0 13Mar19 ??         1:20.37 build/goplugin-tf-aws
```

That's also our suspect for this identity service problem:
```
2019-03-28T18:44:08.487Z [DEBUG] lyra.identity: 2019-03-28T18:44:08.487Z [ERROR] Invoke failed: error="Go function GetExternal returned error 'timeout' (file: /Users/scott/go/pkg/mod/github.com/lyraproj/pcore@v0.0.0-20190320221415-6e1d1bd9c47d/types/function.go, line: 104)"
2019-03-28T18:44:08.497Z [DEBUG] lyra: GetExternal: intId=lyra://puppet.com/aws_ts/aws_subnet1?hid=TerraformAws%3A%3AAws_subnet&rt=TerraformAws%3A%3AAws_subnet extId=<nil>
2019-03-28T18:44:08.499Z [DEBUG] lyra: Create state: intId=lyra://puppet.com/aws_ts/aws_subnet1?hid=TerraformAws%3A%3AAws_subnet&rt=TerraformAws%3A%3AAws_subnet
```
which leads to:
```
invocation of Identity::Service getExternal failed: PCORE_GO_FUNCTION_ERROR Go function GetExternal returned error 'timeout' (file: /Users/scott/go/pkg/mod/github.com/lyraproj/pcore@v0.0.0-20190320221415-6e1d1bd9c47d/types/function.go, line: 104)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/servicesdk@v0.0.0-20190320113120-6be298b18bf0/grpc/client.go:63 (github.com/lyraproj/servicesdk/grpc.(*Client).Invoke)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/wfe@v0.0.0-20190321125752-37a17f3018a7/service/identity.go:33 (github.com/lyraproj/wfe/service.(*identity).getExternal)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/wfe@v0.0.0-20190321125752-37a17f3018a7/service/crud.go:64 (github.com/lyraproj/wfe/service.ApplyState)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/wfe@v0.0.0-20190321125752-37a17f3018a7/wfe/resource.go:64 (github.com/lyraproj/wfe/wfe.(*resource).Run)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/wfe@v0.0.0-20190321125752-37a17f3018a7/wfe/workflowengine.go:453 (github.com/lyraproj/wfe/wfe.(*workflowEngine).runActivity)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/wfe@v0.0.0-20190321125752-37a17f3018a7/wfe/workflowengine.go:403 (github.com/lyraproj/wfe/wfe.(*workflowEngine).activityWorker)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/wfe@v0.0.0-20190321125752-37a17f3018a7/wfe/workflowengine.go:319 (github.com/lyraproj/wfe/wfe.(*workflowEngine).Run.func2)
 at /Users/scott/go/pkg/mod/github.com/lyraproj/pcore@v0.0.0-20190320221415-6e1d1bd9c47d/px/context.go:165 (github.com/lyraproj/pcore/px.Fork.func1)
2019-03-28T18:44:10.583Z [DEBUG] lyra: plugin process exited: path=build/goplugins/identity pid=5950
2019-03-28T18:44:10.583Z [DEBUG] lyra: plugin exited
```

--- a/cmd/lyra/cmd/generate.go
+++ b/cmd/lyra/cmd/generate.go
@@ -36,12 +36,9 @@ func NewGenerateCmd() *cobra.Command {
 }
 
 func runGenerateCmd(cmd *cobra.Command, args []string) {
-	language := args[0]
-	err := generate.Generate(language, targetDirectory)
-	if err != nil {
-		ui.Message("error", err)
-		os.Exit(1)
+	exitCode := generate.Generate(args[0], targetDirectory)
+	if exitCode == 0 {
+		ui.ShowMessage("Generation complete")
 	}
-	ui.ShowMessage("Generation complete")
-	os.Exit(0)
+	os.Exit(exitCode)
 }

--- a/pkg/apply/apply.go
+++ b/pkg/apply/apply.go
@@ -5,13 +5,11 @@ import (
 	"context"
 	"fmt"
 	"os"
-	"os/signal"
 	"strings"
-	"syscall"
+
+	"github.com/lyraproj/lyra/pkg/util"
 
 	"github.com/lyraproj/pcore/utils"
-
-	"github.com/hashicorp/go-plugin"
 
 	"github.com/hashicorp/go-hclog"
 	"github.com/lyraproj/hiera/lookup"
@@ -35,12 +33,6 @@ import (
 type Applicator struct {
 	HomeDir   string
 	DlvConfig string
-}
-
-type cmdError string
-
-func (e cmdError) Error() string {
-	return string(e)
 }
 
 // ApplyWorkflowWithHieraData will apply the named workflow with the supplied hiera data
@@ -99,34 +91,10 @@ func (a *Applicator) ApplyWorkflow(workflowName, hieraDataFilename string, inten
 		`path`:                      types.WrapString(hieraDataFilename),
 		provider.LookupProvidersKey: types.WrapRuntime([]lookup.LookupKey{provider.Yaml, provider.Environment})}
 
-	sgs := make(chan os.Signal, 1)
-	done := make(chan bool, 1)
-
-	// Spawn signal handler routine. It will get called explicitly by the deferred func
-	// below this one unless it is called when a signal is trapped.
-	go func() {
-		<-sgs
-		plugin.CleanupClients()
-		logger.Get().Debug("all plugins cleaned up")
-		done <- true
-	}()
-	signal.Notify(sgs, syscall.SIGINT, syscall.SIGTERM)
-
-	defer func() {
-		if e := recover(); e != nil {
-			exitCode = 1
-			if err, ok := e.(cmdError); ok {
-				ui.Message("error", err)
-			} else {
-				ui.Message("fatal", e)
-			}
-		}
-		sgs <- syscall.SIGUSR1 // Our own
-		<-done
-	}()
-
-	lookup.DoWithParent(context.Background(), provider.MuxLookup, lookupOptions, a.applyWithContext(workflowName, intent))
-	return 0
+	return util.RunCommand(func() int {
+		lookup.DoWithParent(context.Background(), provider.MuxLookup, lookupOptions, a.applyWithContext(workflowName, intent))
+		return 0
+	})
 }
 
 func (a *Applicator) applyWithContext(workflowName string, intent wf.Operation) func(px.Context) {
@@ -165,7 +133,7 @@ func (a *Applicator) parseDlvConfig(c px.Context) {
 	}
 	dc, err := types.Parse(cfg)
 	if err != nil {
-		panic(cmdError(fmt.Sprintf("Unable to parse --dlv option '%s': %s", cfg, err.Error())))
+		panic(util.CmdError(fmt.Sprintf("Unable to parse --dlv option '%s': %s", cfg, err.Error())))
 	}
 	// Pass DlvConfig on to the plugin loader
 	c.Set(api.LyraDlvConfigKey, dc)
@@ -174,7 +142,7 @@ func (a *Applicator) parseDlvConfig(c px.Context) {
 func loadActivity(c px.Context, activityID string) api.Activity {
 	def, ok := px.Load(c, px.NewTypedName(px.NsDefinition, activityID))
 	if !ok {
-		panic(cmdError(fmt.Sprintf("Unable to find definition for activity %s", activityID)))
+		panic(util.CmdError(fmt.Sprintf("Unable to find definition for activity %s", activityID)))
 	}
 	return wfe.CreateActivity(c, def.(serviceapi.Definition))
 }

--- a/pkg/generate/generate.go
+++ b/pkg/generate/generate.go
@@ -2,6 +2,7 @@ package generate
 
 import (
 	"github.com/lyraproj/lyra/pkg/loader"
+	"github.com/lyraproj/lyra/pkg/util"
 	"github.com/lyraproj/pcore/pcore"
 	"github.com/lyraproj/pcore/px"
 	"github.com/lyraproj/servicesdk/lang/typegen"
@@ -11,30 +12,31 @@ import (
 // Generate generates typeset files in the given language for all types exported from known services
 // into the given targetDirectory. If the targetDirectory is the empty string, it will default to
 // plugins/types relative to the current working directory.
-func Generate(language, targetDirectory string) error {
-	pcore.Do(func(c px.Context) {
-		generator := typegen.GetGenerator(language)
+func Generate(language, targetDirectory string) int {
+	return util.RunCommand(func() int {
+		pcore.Do(func(c px.Context) {
+			generator := typegen.GetGenerator(language)
 
-		c.DoWithLoader(loader.New(c.Loader()), func() {
-			loader.LoadPlugins(c)
+			c.DoWithLoader(loader.New(c.Loader()), func() {
+				loader.LoadPlugins(c)
 
-			sNames := c.Loader().Discover(c, func(tn px.TypedName) bool {
-				return tn.Namespace() == px.NsService
-			})
+				sNames := c.Loader().Discover(c, func(tn px.TypedName) bool {
+					return tn.Namespace() == px.NsService
+				})
 
-			if targetDirectory == `` {
-				targetDirectory = "types"
-			}
-			for _, sName := range sNames {
-				if v, ok := px.Load(c, sName); ok {
-					typeSet, _ := v.(serviceapi.Service).Metadata(c)
-					if typeSet != nil && typeSet.Types().Len() > 0 {
-						generator.GenerateTypes(typeSet, targetDirectory)
+				if targetDirectory == `` {
+					targetDirectory = "types"
+				}
+				for _, sName := range sNames {
+					if v, ok := px.Load(c, sName); ok {
+						typeSet, _ := v.(serviceapi.Service).Metadata(c)
+						if typeSet != nil && typeSet.Types().Len() > 0 {
+							generator.GenerateTypes(typeSet, targetDirectory)
+						}
 					}
 				}
-			}
+			})
 		})
+		return 0
 	})
-
-	return nil
 }

--- a/pkg/util/command.go
+++ b/pkg/util/command.go
@@ -1,0 +1,54 @@
+package util
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/lyraproj/lyra/cmd/lyra/ui"
+	"github.com/lyraproj/lyra/pkg/logger"
+)
+
+type CmdError string
+
+func (e CmdError) Error() string {
+	return string(e)
+}
+
+// RunCommand calls the given cmdFunc returns its exit value. It ensures that:
+//
+// A CmdError is logged using an ui.Message (changes exitCode to 1)
+// SIGINT and SIGTERM are handled (changes exitCode to 1)
+// All started plugins are cleaned up
+func RunCommand(cmdFunc func() int) (exitCode int) {
+	sgs := make(chan os.Signal, 1)
+	done := make(chan bool, 1)
+
+	// Spawn signal handler routine. It will get called explicitly by the deferred func
+	// below this one unless it is called when a signal is trapped.
+	go func() {
+		sig := <-sgs
+		plugin.CleanupClients()
+		logger.Get().Debug("all plugins cleaned up")
+		if sig != syscall.SIGUSR1 {
+			exitCode = 1
+		}
+		done <- true
+	}()
+	signal.Notify(sgs, syscall.SIGINT, syscall.SIGTERM)
+
+	defer func() {
+		if e := recover(); e != nil {
+			exitCode = 1
+			if err, ok := e.(CmdError); ok {
+				ui.Message("error", err)
+			} else {
+				ui.Message("fatal", e)
+			}
+		}
+		sgs <- syscall.SIGUSR1 // Our own
+		<-done
+	}()
+	return cmdFunc()
+}

